rasa/core/agent.py

```diff
@@ -112,59 +112,53 @@ async def _pull_model_and_fingerprint(
-                    return None
-                elif resp.status != 200:
-                    logger.debug(
-                        "Tried to fetch model from server, but server response "
-                        "status code is {}. We'll retry later..."
-                        "".format(resp.status)
-                    )
-                    return None
-
-                model_path = Path(model_directory) / resp.headers.get(
-                    "filename", "model.tar.gz"
+    try:
+        params = model_server.combine_parameters()
+        async with model_server.session().request(
+            "GET",
+            model_server.url,
+            timeout=DEFAULT_REQUEST_TIMEOUT,
+            headers=headers,
+            params=params,
+        ) as resp:
+            if resp.status in [204, 304]:
+                logger.debug(
+                    "Model server returned {} status code, "
+                    "indicating that no new model is available. "
+                    "Current fingerprint: {}"
+                    "".format(resp.status, fingerprint)
                 )
-                with open(model_path, "wb") as file:
-                    file.write(await resp.read())
-
-                logger.debug("Saved model to '{}'".format(os.path.abspath(model_path)))
-
-                # return the new fingerprint
-                return resp.headers.get("ETag")
-
-        except aiohttp.ClientError as e:
-            logger.debug(
-                "Tried to fetch model from server, but "
-                "couldn't reach server. We'll retry later... "
-                "Error: {}.".format(e)
+                return None
+            elif resp.status == 404:
+                logger.debug(
+                    "Model server could not find a model at the requested "
+                    "endpoint '{}'. It's possible that no model has been "
+                    "trained, or that the requested tag hasn't been "
+                    "assigned.".format(model_server.url)
+                )
+                return None
+            elif resp.status != 200:
+                logger.debug(
+                    "Tried to fetch model from server, but server response "
+                    "status code is {}. We'll retry later..."
+                    "".format(resp.status)
+                )
+                return None
+            model_path = Path(model_directory) / resp.headers.get(
+                "filename", "model.tar.gz"
             )
-            return None
+            with open(model_path, "wb") as file:
+                file.write(await resp.read())
+            logger.debug("Saved model to '{}'".format(os.path.abspath(model_path)))
+            # return the new fingerprint
+            return resp.headers.get("ETag")
+    except aiohttp.ClientError as e:
+        logger.debug(
+            "Tried to fetch model from server, but "
+            "couldn't reach server. We'll retry later... "
+            "Error: {}.".format(e)
+        )
```

Comment:
Shouldn't we do a `logger.error` here? Same for the non-200 response above

Reply:
I would argue these changes are out of scope for this PR, since I only removed a context manager and copy pasted all the rest of the code.

Reply:
Cool, fair call. Marking this as resolved